src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -12844,15 +13456,42 @@ int8_t Assembler::get_prefixq(Address adr, Register src) {
   return prfx;
 }
 
+int Assembler::get_prefixq_rex2(Address adr, Register src, bool isPage1) {
+  int bits = REXBIT_W;
+  if (isPage1) bits |= REX2BIT_M0;
+  if (adr.base()->encoding() >= 24) bits |= (REX2BIT_B4 | REXBIT_B);
+  else if (adr.base()->encoding() >= 16) bits |= REX2BIT_B4;
+  else if (adr.base()->encoding() >= 8) bits |= REXBIT_B;
+  if (adr.index()->encoding() >= 24) bits |= (REX2BIT_X4 | REXBIT_X);
+  else if (adr.index()->encoding() >= 16) bits |= REX2BIT_X4;
+  else if (adr.index()->encoding() >= 8) bits |= REXBIT_X;
+  if (src->encoding() >= 24) bits |= (REX2BIT_R4 | REXBIT_R);
+  else if (src->encoding() >= 16) bits |= REX2BIT_R4;
+  else if (src->encoding() >= 8) bits |= REXBIT_R;
```

Comment:
We could have 3 basic functions:
   get_src_prefix_bits(int enc)  
   get_base_or_dst_prefix_bits(int enc)
   get_index_prefix_bitx(int enc)
get_src_prefix_bits could have logic something like below:
   int bits = 0;
   if ((enc & 8) == 8) bits |= REXBIT_R;
   if ((enc & 16) == 16) bits |= REXBIT_R4;
   return bits;

get_base_or_dst_prefix_bits could have logic something like below:
   int bits = 0;
   if ((enc & 8) == 8) bits |= REXBIT_B;
   if ((enc & 16) == 16) bits |= REXBIT_B4;
   return bits;
get_index_prefix_bits could have logic something like below:
   int bits = 0;
   if ((enc & 8) == 8) bits |= REXBIT_X;
   if ((enc & 16) == 16) bits |= REXBIT_X4;
   return bits;
These could then be used across all the prefix encoding methods.
get_prefixq_rex2 selected lines then become:  
  bits |= get_base_or_dst_prefix_bits(adr.base()->encoding());
  bits |= get_index_prefix_bits(adr.index()->encoding());
  bits |= get_src_prefix_bits(src->encoding());

Reply:
Done. Thanks.